### ISSUE TITLE
Major Difference in the output of Torch Model and Onnx Model for Zipformer2

### ISSUE THREAD

**Issue by bhaswa** [k2-fsa/icefall#1388]
Hi,

I have trained zipformer2 (without streaming) model with my dataset.

Training command:  **./zipformer/train.py --num-epochs 40 --start-epoch 1 --use-fp16 1 --enable-musan False --exp-dir zipformer/exp-small --causal 0 --num-encoder-layers 2,2,2,2,2,2 --feedforward-dim 512,768,768,768,768,768 --encoder-dim 192,256,256,256,256,256 --encoder-unmasked-dim 192,192,192,192,192,192 --base-lr 0.04 --full-libri 1** 

When I decode from the .pt model, the output is nearly same as the GT

Decoding Command (for .pt Model): **./zipformer/decode.py --epoch 40 --use-averaged-model False --avg 1 --exp-dir zipformer/exp-small/ --decoding-method greedy_search --causal 0 --num-encoder-layers 2,2,2,2,2,2 --feedforward-dim 512,768,768,768,768,768 --encoder-dim 192,256,256,256,256,256 --encoder-unmasked-dim 192,192,192,192,192,192 --chunk-size 16 --left-context-frames 64 --bpe-model data/lang_bpe_500/bpe.model --manifest-dir data/test/fbank/**

But when I convert the model in onnx and decode using the onnx model, most of the segments are missing in the decoded output.

Onnx Conversion Commnad:  **./zipformer/export-onnx-bhaswati.py --tokens data/lang_bpe_500/tokens.txt --use-averaged-model 1 --epoch 40 --avg 1 --exp-dir zipformer/exp-small/ --num-encoder-layers "2,2,2,2,2,2" --feedforward-dim "512,768,768,768,768,768" --encoder-dim "192,256,256,256,256,256" --encoder-unmasked-dim "192,192,192,192,192,192" --causal False --chunk-size 16 --left-context-frames 64**

Decoding Command (for .onnx Model): **./zipformer/onnx_pretrained.py --encoder-model-filename zipformer/exp-small/encoder-epoch-40-avg-1.int8.onnx --decoder-model-filename zipformer/exp-small/decoder-epoch-40-avg-1.int8.onnx --joiner-model-filename [zipformer/exp-small/joiner-epoch-40-avg-1.int8.onnx --tokens data/lang_bpe_500/tokens.txt audio.wav**

**Comment by csukuangfj** [comment]
> Decoding Command (for .pt Model): ./zipformer/decode.py --epoch 40 --use-averaged-model False 

> Onnx Conversion Commnad: ./zipformer/export-onnx-bhaswati.py --tokens data/lang_bpe_500/tokens.txt --use-averaged-model 1 

Could you use the same value for `--use-averaged-model 1 ` and then test it again?

**Comment by bhaswa** [comment]
The model behaves the same way with --use-averaged-model 1

**Comment by csukuangfj** [comment]
> most of the segments are missing in the decoded output.

Could you give a detailed example out?

**Comment by bhaswa** [comment]
I have uploaded the model and the audio [here](https://we.tl/t-s88Q1bzwOU), if it helps.

Below are the two outputs from pth model and onnx model. 

Onnx Conversion Commnad: ./zipformer/export-onnx-bhaswati.py --tokens data/lang_bpe_500/tokens.txt --use-averaged-model 1 --epoch 40 --avg 1 --exp-dir zipformer/exp-small/ --num-encoder-layers "2,2,2,2,2,2" --feedforward-dim "512,768,768,768,768,768" --encoder-dim "192,256,256,256,256,256" --encoder-unmasked-dim "192,192,192,192,192,192" --causal False --chunk-size 16 --left-context-frames 64

Decoding Command (for .onnx Model): ./zipformer/onnx_pretrained.py --encoder-model-filename zipformer/exp-small/encoder-epoch-40-avg-1.int8.onnx --decoder-model-filename zipformer/exp-small/decoder-epoch-40-avg-1.int8.onnx --joiner-model-filename [zipformer/exp-small/joiner-epoch-40-avg-1.int8.onnx --tokens data/lang_bpe_500/tokens.txt audio.wav

Seems like there is some issue while converting the model to onnx. Majority of the portion of the audio is missing in onnx decode result.

**Comment by csukuangfj** [comment]
> I have uploaded the model and the audio [here](https://we.tl/t-s88Q1bzwOU), if it helps.

Could you post some text results? The files in your link are > 100MB, which is too large for me to download.

Also, could you test not using `.int8.onnx`?

**Comment by csukuangfj** [comment]
Could you also post the logs of `export-onnx-bhaswati.py` and `decode.py`? Only the first part of the logs that contains detailed parameters is enough.

**Comment by bhaswa** [comment]
Did you get any chance to look into the logs ?

**Comment by csukuangfj** [comment]
> Did you get any chance to look into the logs ?

The logs look correct.

Does pretrained.py give correct result?

By default decode.py uses pre-computed features while onnx_pretrained.py and pretrained.py compute features on the fly.

**Comment by bhaswa** [comment]
Output of pretrained.py:

θBΩə θF μFOCSQə βMλə OBλə C∞ψθLλJ∞ζə ζəλə θBΩə BC ΩK@ə γC μFOCSQə C∞βJO!ə ΩK@ə CSOφBθəλ@ə βθF @Jρə γKOə L∞əμC K∞@ə BC βλMεə θLλə J∞@ə !F OMμə εC γKOə γə!ə ψC @C@ə∞!ə !əλ∞ə Mβə ζMλC Kεə BC M@Cγəμə ζəλə ΩKμL θBΩə OK∞ə θF ΩCθəλə εC

The output is almost similar to decode.py

The problem of decoded result getting blanked out is not happening here.

**Comment by bhaswa** [comment]
Did you get any chance to look what is the issue here ?
